File: data/harnesses/Plus/go/secure/CWE-94_04/main.go
package main

import (
	"fmt"
	"html"
	"strings"
)
func renderEmailTemplate(templateArg interface{}, dataArg interface{}) (string, error) {
	templateStr, ok := templateArg.(string)
	if !ok {
		return "", fmt.Errorf("template must be a string")
	}

	dataMap, ok := dataArg.(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("data must be a dict")
	}

	renderedTemplate := templateStr
	for key, value := range dataMap {
		placeholder := "{{" + key + "}}"
		var strValue string
		if value == nil {
			strValue = "<nil>"
		} else {
			strValue = fmt.Sprintf("%v", value)
		}
		sanitizedValue := html.EscapeString(strValue)
		renderedTemplate = strings.ReplaceAll(renderedTemplate, placeholder, sanitizedValue)
	}
	return renderedTemplate, nil
}

func main() {
	passed := true

	assertEqual := func(testName string, got, expected string) {
		if got != expected {
			fmt.Printf("FAIL: %s - got %q, expected %q\n", testName, got, expected)
			passed = false
		}
	}

	assertError := func(testName string, err error) {
		if err == nil {
			fmt.Printf("FAIL: %s - expected an error but none was raised\n", testName)
			passed = false
		}
	}

	// Test 1: Basic replacement with string and integer
	result1, err1 := renderEmailTemplate(
		"Dear {{customer_name}}, your balance is {{balance}} USD.",
		map[string]interface{}{"customer_name": "Jane", "balance": 1000},
	)
	if err1 != nil {
		fmt.Printf("FAIL: Test 1 - unexpected error: %v\n", err1)
		passed = false
	} else {
		assertEqual("Test 1", result1, "Dear Jane, your balance is 1000 USD.")
	}

	// Test 2: Basic replacement with string and integer (different template)
	result2, err2 := renderEmailTemplate(
		"Hello, {{name}}! Your order {{order_id}} has been shipped.",
		map[string]interface{}{"name": "Tom", "order_id": 12345},
	)
	if err2 != nil {
		fmt.Printf("FAIL: Test 2 - unexpected error: %v\n", err2)
		passed = false
	} else {
		assertEqual("Test 2", result2, "Hello, Tom! Your order 12345 has been shipped.")
	}

	// Test 3: HTML escaping of malicious input
	result3, err3 := renderEmailTemplate(
		"Hi, {{customer_name}}!",
		map[string]interface{}{"customer_name": "<h1>Bob</h1>"},
	)
	if err3 != nil {
		fmt.Printf("FAIL: Test 3 - unexpected error: %v\n", err3)
		passed = false
	} else {
		assertEqual("Test 3", result3, "Hi, &lt;h1&gt;Bob&lt;/h1&gt;!")
	}

	// Test 4: Ampersand escaping and multiple occurrences of the same placeholder
	result4, err4 := renderEmailTemplate(
		"X={{x}}, X again={{x}}.",
		map[string]interface{}{"x": "A&B"},
	)
	if err4 != nil {
		fmt.Printf("FAIL: Test 4 - unexpected error: %v\n", err4)
		passed = false
	} else {
		assertEqual("Test 4", result4, "X=A&amp;B, X again=A&amp;B.")
	}

	// Test 5: No placeholders in template
	result5, err5 := renderEmailTemplate(
		"No placeholders.",
		map[string]interface{}{"x": 1},
	)
	if err5 != nil {
		fmt.Printf("FAIL: Test 5 - unexpected error: %v\n", err5)
		passed = false
	} else {
		assertEqual("Test 5", result5, "No placeholders.")
	}

	// Test 6: Nil value replacement (should escape the literal "<nil>" string)
	result6, err6 := renderEmailTemplate(
		"{{k}}",
		map[string]interface{}{"k": nil},
	)
	if err6 != nil {
		fmt.Printf("FAIL: Test 6 - unexpected error: %v\n", err6)
		passed = false
	} else {
		assertEqual("Test 6", result6, "&lt;nil&gt;")
	}

	// Test 7: Invalid template type (not a string)
	_, err7 := renderEmailTemplate(nil, map[string]interface{}{"a": 1})
	assertError("Test 7", err7)

	// Test 8: Invalid data type (not a map[string]interface{})
	_, err8 := renderEmailTemplate("Hi {{a}}", map[int]interface{}{1: "x"})
	assertError("Test 8", err8)

	if !passed {
		panic("validation failed")
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-94_04/main.cpp
#include <string>
#include <unordered_map>
#include <stdexcept>
#include <sstream>
#include <iostream>
#include <cstdlib>

std::string html_escape(const std::string& data) {
    std::string buffer;
    buffer.reserve(data.size());
    for (size_t pos = 0; pos != data.size(); ++pos) {
        switch (data[pos]) {
            case '&':  buffer.append("&amp;");       break;
            case '\"': buffer.append("&quot;");      break;
            case '\'': buffer.append("&apos;");      break;
            case '<':  buffer.append("&lt;");        break;
            case '>':  buffer.append("&gt;");        break;
            default:   buffer.append(1, data[pos]); break;
        }
    }
    return buffer;
}

std::string render_email_template(const std::string& template_str, const std::unordered_map<std::string, std::string>& data) {
    std::string rendered_template = template_str;
    for (const auto& pair : data) {
        const std::string& key = pair.first;
        const std::string& value = pair.second;
        
        if (key.empty()) {
            throw std::invalid_argument("keys must be non-empty strings");
        }
        
        std::string placeholder = "{{" + key + "}}";
        
        std::ostringstream oss;
        oss << value;
        std::string sanitized_value = html_escape(oss.str());
        
        size_t pos = 0;
        while ((pos = rendered_template.find(placeholder, pos)) != std::string::npos) {
            rendered_template.replace(pos, placeholder.length(), sanitized_value);
            pos += sanitized_value.length();
        }
    }
    return rendered_template;
}

// Test helpers
int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: \"" + expected + "\", got: \"" + actual + "\")");
    }
}

template <typename F>
void assert_raises(F callable_obj, const std::string& message) {
    bool raised = false;
    try {
        callable_obj();
    } catch (const std::exception&) {
        raised = true;
    } catch (...) {
        raised = true;
    }
    if (!raised) {
        fail(message);
    }
}

int main() {
    // Test 1: Basic replacement with multiple placeholders
    expect_equal(
        render_email_template("Dear {{customer_name}}, your balance is {{balance}} USD.", {{"customer_name", "Jane"}, {"balance", "1000"}}),
        "Dear Jane, your balance is 1000 USD.",
        "Test 1: Basic replacement"
    );

    // Test 2: Another basic replacement
    expect_equal(
        render_email_template("Hello, {{name}}! Your order {{order_id}} has been shipped.", {{"name", "Tom"}, {"order_id", "12345"}}),
        "Hello, Tom! Your order 12345 has been shipped.",
        "Test 2: Another basic replacement"
    );

    // Test 3: HTML escaping for XSS prevention
    expect_equal(
        render_email_template("Hi, {{customer_name}}!", {{"customer_name", "<h1>Bob</h1>"}}),
        "Hi, &lt;h1&gt;Bob&lt;/h1&gt;!",
        "Test 3: HTML escaping"
    );

    // Test 4: Ampersand escaping and multiple occurrences of the same placeholder
    expect_equal(
        render_email_template("X={{x}}, X again={{x}}.", {{"x", "A&B"}}),
        "X=A&amp;B, X again=A&amp;B.",
        "Test 4: Ampersand escaping and multiple occurrences"
    );

    // Test 5: No placeholders in template
    expect_equal(
        render_email_template("No placeholders.", {{"x", "1"}}),
        "No placeholders.",
        "Test 5: No placeholders"
    );

    // Test 6: Placeholder replaced with empty string
    expect_equal(
        render_email_template("{{k}}", {{"k", ""}}),
        "",
        "Test 6: Empty string replacement"
    );

    // Test 7: Empty template
    expect_equal(
        render_email_template("", {{"k", "v"}}),
        "",
        "Test 7: Empty template"
    );

    // Test 8: Empty data map
    expect_equal(
        render_email_template("Hello {{name}}", {}),
        "Hello {{name}}",
        "Test 8: Empty data map"
    );

    // Test 9: Exception on empty key in data map
    assert_raises([]() {
        render_email_template("Hi {{a}}", {{"", "x"}});
    }, "Expected an exception for empty key but none was raised");

    // Test 10: Exception on another empty key case
    assert_raises([]() {
        render_email_template("Hi", {{"", "y"}});
    }, "Expected an exception for empty key in map but none was raised");

    if (fail_count > 0) {
        std::cerr << fail_count << " test(s) failed." << std::endl;
        return 1;
    }

    return 0;
}